### PR TITLE
Implement software write watch and make concurrent GC functional outs…

### LINKED ISSUE #2224: Enable Background GC on Linux
Background GC relies on the Write Watch support provided by Windows. There are a few different approaches we can take in order to implement Background/Concurrent GC on Linux. We need to investigate how we can achieve this in the most efficient way.

### PR DESCRIPTION
**Pull request by kouvel**
…ide Windows

Fixes #2224:
- Implemented software write watch using write barriers
- A new set of write barriers is introduced, each corresponding to an existing one, but which also updates the write watch table. The GC switches to a write watch barrier during concurrent GC, and switches back to a non write watch barrier after the final query for dirty pages.
- The write watch table is alloacted along with the card table
- Since the card table is used differently, different synchonization is used for the write watch table. The runtime is suspended during resize since that is the most infrequently occuring operation, of that, ResetWriteWatch, and GetWriteWatch.
- ResetWriteWatch() doesn't need a suspend, but since the software WW version is much faster than the Windows version, moved it into the suspended region to avoid some synchronization that would otherwise be required
- The background calls to GetWriteWatch() don't need or do a suspend. They only need to synchronize with the resize path, not for the purpose of correct functionality, but to not miss dirty pages such that concurrent GC is effective.
- Also, while concurrent GC was already enabled by default, it wasn't actually on by default outside Windows since there was no write watch support. With this change, concurrent GC is on and functions by default.

Miscellaneous:
- Fixed runtests.sh to copy mscorlib.dll and delete the Windows version of mscorlib.ni.dll

### CODE REVIEW

**Comment by kouvel** [comment]
CC @sergiy-k  @Maoni0 @janvorli @jkotas

**Comment by kouvel** [comment]
The change is undergoing stress testing, please don't merge until [WIP] is removed from the title.

**Comment by kouvel** [comment]
@dotnet-bot test Ubuntu pri1

**Comment by kouvel** [comment]
@dotnet-bot test OSX pri1

File: src/gc/env/gcenv.base.h [hunk 1]
```diff
@@ -543,8 +545,8 @@ void LogSpewAlways(const char *fmt, ...);
 
 // -----------------------------------------------------------------------------------------------------------
 
-void StompWriteBarrierEphemeral();
-void StompWriteBarrierResize(bool bReqUpperBoundsCheck);
+void StompWriteBarrierEphemeral(bool isRuntimeSuspended);
+void StompWriteBarrierResize(bool isRuntimeSuspended, BOOL bReqUpperBoundsCheck);
```
Comment (jkotas) [inline src/gc/env/gcenv.base.h]: Nit - could you please make the second argument lowercase bool as well?

File: src/gc/gc.cpp [hunk 2]
```diff
@@ -4291,7 +4318,7 @@ void* virtual_alloc (size_t size)
 
 void virtual_free (void* add, size_t size)
 {
-    GCToOSInterface::VirtualRelease (add, size);
+    GCToOSInterface::VirtualRelease (add);
```
Comment (jkotas) [inline src/gc/gc.cpp]: Revert this change - the size argument is intentional to allow GC to be used in different environments (in particular on Unix version of CoreRT), even though it is not used in CoreCLR.

File: src/vm/amd64/jithelpers_fastwritebarriers.S [hunk 3]
```diff
@@ -244,3 +246,331 @@ PATCH_LABEL JIT_WriteBarrier_SVR64_PatchLabel_CardTable
         mov     byte ptr [rdi + rax], 0FFh
         ret
 LEAF_END_MARKED JIT_WriteBarrier_SVR64, _TEXT
+
+#endif
+
+
+#ifdef FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
+
```
Comment (jkotas) [inline src/vm/amd64/jithelpers_fastwritebarriers.S]: Are we gaining much by having all the different optimized flavors of the write barrier with software write watch? It is pretty non-trivial to test them all...

File: src/gc/sample/gcenv.unix.cpp [hunk 4]
```diff
@@ -338,6 +338,19 @@ ResetWriteWatch(
     return 1;
 }
 
+WINBASEAPI
+UINT
+WINAPI
+ResetWriteWatch(
+    LPVOID lpBaseAddress,
```
Comment (jkotas) [inline src/gc/sample/gcenv.unix.cpp]: Looks like accidental change - undo?

File: src/gc/env/gcenv.os.h [hunk 5]
```diff
@@ -87,7 +87,7 @@ class GCToOSInterface
     //  size    - size of the virtual memory range
     // Return:
     //  true if it has succeeded, false if it has failed
-    static bool VirtualRelease(void *address, size_t size);
+    static bool VirtualRelease(void *address);
```
Comment (janvorli) [inline src/gc/env/gcenv.os.h]: This change is bad. The VirtualRelease must have the size argument. While it is not used on CoreCLR and Windows, it is required for CoreRT on Unix.

File: src/gc/env/gcenv.os.h [hunk 6]
```diff
@@ -138,7 +138,7 @@ class GCToOSInterface
     //                       on output, the number of page addresses that are returned in the array.
     // Return:
     //  true if it has succeeded, false if it has failed
-    static bool GetWriteWatch(bool resetState, void* address, size_t size, void** pageAddresses, uintptr_t* pageAddressesCount);
+    static void GetWriteWatch(bool resetState, void* address, size_t size, void** pageAddresses, uintptr_t* pageAddressesCount);
```
Comment (janvorli) [inline src/gc/env/gcenv.os.h]: What is the reason for changing the return from bool to void?

File: src/vm/amd64/jithelpers_fast.S [hunk 7]
```diff
@@ -10,6 +10,45 @@ LEAF_ENTRY JIT_PatchedCodeStart, _TEXT
         ret
 LEAF_END JIT_PatchedCodeStart, _TEXT
 
+
```
Comment (janvorli) [inline src/vm/amd64/jithelpers_fast.S]: Is there a reason for moving this function here? It seems to just complicate the diff of the change.

File: src/gc/softwarewritewatch.cpp [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+    if (dirtyPageCount == 0)
+    {
+        return;
+    }
+
+    if (!isRuntimeSuspended)
+    {
+        // When a page is marked as dirty, a memory barrier is not issued after the write most of the time. Issue a memory
+        // barrier on all active threads of the process now to make recent changes to dirty state visible to this thread.
+        FlushProcessWriteBuffers();
```
Comment (janvorli) [inline src/gc/softwarewritewatch.cpp]: Please use GCToOSInterface::FlushProcessWriteBuffers to keep the GC code OS independent

File: src/gc/softwarewritewatch.cpp [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+
+    if (!isRuntimeSuspended && clearDirty && dirtyPageIndex != 0)
+    {
+        // When dirtying a page, the dirty state of the page is first checked to see if the page is already dirty. If already
+        // dirty, the write to mark it as dirty is skipped. So, when the dirty state of a page is cleared, we need to make sure
+        // the cleared state is visible to other threads that may dirty the page, before marking through objects in the page, so
+        // that the GC will not miss marking through dirtied objects in the page. Issue a memory barrier on all active threads
+        // of the process now.
+        MemoryBarrier(); // flush writes from this thread first to guarantee ordering
+        FlushProcessWriteBuffers();
```
Comment (janvorli) [inline src/gc/softwarewritewatch.cpp]: Please use GCToOSInterface::FlushProcessWriteBuffers to keep the GC code OS independent

**Comment by kouvel** [comment]
Made all of the changes suggested above

File: src/vm/amd64/jithelpers_fastwritebarriers.S (hunk 3 above)
Comment (kouvel) [inline src/vm/amd64/jithelpers_fastwritebarriers.S]: I'm not sure in what scenarios the different write barriers helped originally when they were written. I assumed there are scenarios where each is useful. The benchmarks I'm running spend relatively little time in write barriers, and I don't see any noticeable perf difference between PreGrow64 and PostGrow64. I don't see the 32-bit constant write barriers being used at all normally, but not sure if there's a scenario where they would be used, and how much better perf they would give.

File: src/gc/env/gcenv.os.h (hunk 6 above)
Comment (kouvel) [inline src/gc/env/gcenv.os.h]: All of the callers were just asserting that the return value is true, so I don't see any reason for the return value. Anyway, this is not important to my change, reverted.

File: src/vm/amd64/jithelpers_fast.S (hunk 7 above)
Comment (kouvel) [inline src/vm/amd64/jithelpers_fast.S]: Due to the relative jump from CheckedWriteBarrier to WriteBarrier, the negative jump overflowed one byte with WriteBarrier increasing in size. Moving it before makes the jump distance independent of both functions' sizes. This order is also consistent with the windows assembly file, making it easier to diff those two.

Comment (janvorli) [inline src/vm/amd64/jithelpers_fast.S]: Ok, makes sense then.

File: src/vm/amd64/jithelpers_fastwritebarriers.S (hunk 3 above)
Comment (jkotas) [inline src/vm/amd64/jithelpers_fastwritebarriers.S]: These were written 10+ years ago in quite different world, to optimize for early versions of x64 hardware: 
- There was no address space randomization, so you had a very good chance of getting addresses in low 2GB - not the case today.
- An extra instruction on a hot path mattered a lot more than how much it matters today.

We need to make sure that we have test coverage for all the variants with the software write watch. E.g. if you do not see the 32-bit constant ones getting hit in tests - we should not have them, so that they won't get hit by accident in production.

File: src/vm/arm/stubs.cpp [hunk 10]
```diff
@@ -469,7 +469,7 @@ void StompWriteBarrierResize(BOOL bReqUpperBoundsCheck)
     GCStressPolicy::InhibitHolder iholder;
 
     bool fSuspended = false;
-    if (!g_fEEInit && !GCHeap::IsGCInProgress())
+    if (!g_fEEInit && !GCHeap::IsGCInProgress() && !isRuntimeSuspended)
```
Comment (jkotas) [inline src/vm/arm/stubs.cpp]: Can this be just `if (!isRuntimeSuspended)` now?

File: src/vm/gchelpers.cpp [hunk 11]
```diff
@@ -1319,6 +1341,13 @@ void ErectWriteBarrierForMT(MethodTable **dst, MethodTable *ref)
     
     if (ref->Collectible())
     {
+#ifdef FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
+        if (SoftwareWriteWatch::IsEnabledForGCHeap())
+        {
```
Comment (jkotas) [inline src/vm/gchelpers.cpp]: Is this one actually needed for anything?

File: src/vm/amd64/jithelpers_fastwritebarriers.S (hunk 3 above)
Comment (Maoni0) [inline src/vm/amd64/jithelpers_fastwritebarriers.S]: I don't think we would see the 32-bit ones getting hit in our tests because we have the high entropy flag on which enables ASLR to reserve the lower addresses. The question is more that if we used the 64-bit ones to handle the 32-bit addresses, is that a big perf loss. I doubt it. As far as the pre and post 64 ones, it's less clear to me - I would think that having pre could give us a slight perf boost in workstation GC (that's hard to detect with microbenchmarks). @PeterSolMS , do you have an opinion?

Comment (kouvel) [inline src/vm/amd64/jithelpers_fastwritebarriers.S]: I don't think there will be much perf difference between the 32-bit and 64-bit versions either. We probably don't even gain any perf from patching the constants, as I have noticed in the past that Intel processors are pretty good at prefetching constant loads, making their cost negligible most of the time (depending on register use).

For testing the 32-bit barriers, I was planning on verifying them in the debugger. These are small functions with only a handful of paths, I figured it would be adequate to verify all paths through a debugger, as their behavior is not going to change beyond that.

Also to clarify @jkotas, are you suggesting to remove only the write watch versions of the 32-bit constant barriers, or to remove all 32-bit constant barriers?

Comment (jkotas) [inline src/vm/amd64/jithelpers_fastwritebarriers.S]: I would remove all 32-bit constant ones - while you are on it.

File: src/vm/gchelpers.cpp (hunk 11 above)
Comment (kouvel) [inline src/vm/gchelpers.cpp]: My understanding is that this only applies to objects from collectible assemblies, and doesn't apply to CoreCLR. I don't know what the significance of the code here is, but I just added the update here for completeness, as it looks like a place where the write watch table would need to be updated if the feature applies.

File: src/vm/arm/stubs.cpp (hunk 10 above)
Comment (kouvel) [inline src/vm/arm/stubs.cpp]: Yep will fix this as a few other similar checks

**Comment by kouvel** [comment]
Fixed issues mentioned above and removed 32-bit constant write barriers on amd64, please see last two commits

**Comment by kouvel** [comment]
@dotnet-bot test this please

File: src/vm/gchelpers.cpp (hunk 11 above)
Comment (jkotas) [inline src/vm/gchelpers.cpp]: Collectible assemblies (created via Reflection.Emit) are exposed by CoreCLR.

Comment (kouvel) [inline src/vm/gchelpers.cpp]: Sure, I'll ask around and try to get some context on what the scenarios are for this. It may be suboptimal if unnecessary, but it shouldn't be functionally incorrect to mark more pages as dirty, so it should be a safe change in that regard.

**Comment by kouvel** [comment]
@dotnet-bot test Windows_NT corefx_baseline

**Comment by kouvel** [comment]
@dotnet-bot test Ubuntu corefx_baseline

**Comment by kouvel** [comment]
@dotnet-bot test OSX corefx_baseline

**Comment by kouvel** [comment]
@dotnet-bot test Windows_NT corefx_baseline

**Comment by kouvel** [comment]
@dotnet-bot test OSX corefx_baseline

**Comment by kouvel** [comment]
@dotnet-bot test this please

**Comment by kouvel** [comment]
@dotnet-bot test this please

**Comment by kouvel** [comment]
@dotnet-bot test CentOS7.1 Debug

**Comment by kouvel** [comment]
@dotnet-bot test this please

**Comment by kouvel** [comment]
The Ubuntu CoreFx test failures seem to be unrelated to this change.

**Comment by kouvel** [comment]
Rebased away other commits. Testing is complete for this change, removed [WIP] from title, and it's ready to go after the tests here finish.

**Comment by sdmaclea** [comment]
@kouvel @swgillespie @jkotas Is there any reason the Windows *.asm files were changed?  Since this is intended for Unix it seems like it should not have been necessary.  It would seem it would be untested.

**Comment by kouvel** [comment]
I believe the Windows asm code was only changed under the macro FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP, which is disabled by default on Windows. It was to make it easy to enable on Windows for perf comparisons.
